### PR TITLE
fix for different output from 'luksAddKey' command w/cryptsetup v2.0.2

### PR DESCRIPTION
**Pull request by therealjuanmartinez**
Below is the output of the `cryptsetup luksAddKey` command on Ubuntu 18.04 running cryptsetup v2.0.2:

`$ cryptsetup --version
cryptsetup 2.0.2`

`$ sudo cryptsetup luksAddKey -v /dev/vg1/lv1  < <(echo "myPassphrase"; echo -n "newKey")`
`Key slot 0 unlocked.`
`Key slot 0 unlocked.`
`Command successful.`

My change accommodates the use of '`unlocked`' vs, '`created`' as well as the fact that two identical lines are output.  
Additionally, errors can occur at line 142 without being passed on, therefore, I've added a 2nd if/then catch for a case when `SLT` does not get properly assigned a value by this point.

### CODE REVIEW

File: src/luks/clevis-luks-bind.in [hunk 1]
```diff
@@ -139,14 +139,19 @@ esac
 if [ -n "$SLT" ]; then
     cryptsetup luksAddKey --key-slot "$SLT" "$DEV"
 else
-    SLT="$(cryptsetup luksAddKey -v "$DEV" \
-        | sed -rn 's|^Key slot ([0-9]+) created\.$|\1|p')"
-fi < <(echo "$old"; echo -n "$key")
+    SLT=$(cryptsetup luksAddKey -v "$DEV" \
+	    | sed -rn 's|^Key slot ([0-9]+) (unlocked\|created)\.$|\1|p' | head -1)
```
[the remark this answers is not in the document]
Comment (sergio-correia) [reply src/luks/clevis-luks-bind.in]: By the way, I was testing something like what I mentioned in here: https://github.com/sergio-correia/clevis/commit/e1c4bc948dc8b8b7b9fc162ed77fd636467a4eb6

**Comment by sergio-correia** [comment]
@therealjuanmartinez: will you update this PR or would you prefer me to submit a PR with the fix I suggested?

**Comment by therealjuanmartinez** [comment]
I agree there ideally should not be a dependency on parsing locale-specific output

File: src/luks/clevis-luks-bind.in [hunk 2]
```diff
@@ -135,18 +136,47 @@ case "$KEY" in
  *) old="$(/bin/cat "$KEY")";;
 esac
 
-# Add the new key
+
+if [ -z $SLT ]; then
```
Comment (sergio-correia) [inline src/luks/clevis-luks-bind.in]: Please, move this full block to right before the `cryptsetup luksAddKey` inside the `else`, further down. No need to test for `$SLT` being empty there, since we already know that is true.

File: src/luks/clevis-luks-bind.in [hunk 3]
```diff
@@ -135,18 +136,47 @@ case "$KEY" in
...
+    if [ luks_type == "luks2" ]; then
```
Comment (sergio-correia) [inline src/luks/clevis-luks-bind.in]: Typo when checking the variable luks_type: `"$luks_type"`

File: src/luks/clevis-luks-bind.in [hunk 4]
```diff
@@ -135,18 +136,47 @@ case "$KEY" in
...
+            readarray -t usedSlotsBeforeAddKey < <(cryptsetup luksDump "${DEV}" | sed -rn 's|^\s+([0-9]+): luks2$|\1|p')
```
Comment (sergio-correia) [inline src/luks/clevis-luks-bind.in]: The indent in both these `readarray` lines seems off. Please, adjust it and also break the lines before the pipe, so that it does not get too long.

File: src/luks/clevis-luks-bind.in [hunk 5]
```diff
@@ -135,18 +136,47 @@ case "$KEY" in
...
+fi
+
+#Add the new key
 if [ -n "$SLT" ]; then
     cryptsetup luksAddKey --key-slot "$SLT" "$DEV"
 else
-    SLT="$(cryptsetup luksAddKey -v "$DEV" \
-        | sed -rn 's|^Key slot ([0-9]+) created\.$|\1|p')"
-fi < <(echo "$old"; echo -n "$key")
+    cryptsetup luksAddKey -v "$DEV"
```
Comment (sergio-correia) [inline src/luks/clevis-luks-bind.in]: We can remove the `-v` here. By the way, please rebase this branch on top of the current master. There were a few changes in this specific area, as now we are using key files to pass the password to cryptsetup.

File: src/luks/clevis-luks-bind.in [hunk 6]
```diff
@@ -135,18 +136,47 @@ case "$KEY" in
...
+    cryptsetup luksAddKey -v "$DEV" 
+fi < <(echo "$old"; echo -n "$key") 
 if [ $? -ne 0 ]; then
     echo "Error while adding new key to LUKS header!" >&2
     exit 1
 fi
 
+#Find slot used if not defined previously
+if [ -z $SLT ]; then
+    if [ luks_type == "luks2" ]; then
```
Comment (sergio-correia) [inline src/luks/clevis-luks-bind.in]: Typo with luks_type variable: should be `"$luks_type"`

File: src/luks/clevis-luks-bind.in [hunk 7]
```diff
@@ -135,18 +136,47 @@ case "$KEY" in
...
+            readarray -t usedSlotsAfterAddKey < <(cryptsetup luksDump "${DEV}" | sed -rn 's|^\s+([0-9]+): luks2$|\1|p')
+    else
+            readarray -t usedSlotsAfterAddKey < <(cryptsetup luksDump "${DEV}" | sed -rn 's|^Key Slot ([0-7]): ENABLED$|\1|p')
```
Comment (sergio-correia) [inline src/luks/clevis-luks-bind.in]: Like before, please adjust the indent and break the long line.

File: src/luks/clevis-luks-bind.in [hunk 8]
```diff
@@ -135,18 +136,47 @@ case "$KEY" in
...
+    fi
+    for i in "${usedSlotsAfterAddKey[@]}"; do
+            if [[ ! " ${usedSlotsBeforeAddKey[@]} " =~ " ${i} " ]]; then
+                    SLT=$i
+                    break
+            fi
+    done
```
Comment (sergio-correia) [inline src/luks/clevis-luks-bind.in]: Please adjust the indent in this block as well.

File: src/luks/clevis-luks-bind.in [hunk 9]
```diff
@@ -135,18 +136,47 @@ case "$KEY" in
...
+fi
+
+if [ -z $SLT ]; then
+	echo "Error while adding new key to LUKS header! Key slot is undefined." >&2
```
Comment (sergio-correia) [inline src/luks/clevis-luks-bind.in]: Please adjust the indent here as well. It seems you are now using tabs in this point.

File: src/luks/clevis-luks-bind.in [hunk 10]
```diff
@@ -135,20 +136,48 @@ case "$KEY" in
...
+
+#Add the new key
 if [ -n "$SLT" ]; then
     cryptsetup luksAddKey --key-slot "$SLT" --key-file \
         <(echo -n "$existing_key") "$DEV"
 else
-    SLT="$(cryptsetup luksAddKey \
-        --key-file <(echo -n "${existing_key}") -v "$DEV" \
-        | sed -rn 's|^Key slot ([0-9]+) created\.$|\1|p')"
+    if [ $luks_type == "luks2" ]; then
```
Comment (sergio-correia) [inline src/luks/clevis-luks-bind.in]: Please, double-quote `"$luks_type"` here.

File: src/luks/clevis-luks-bind.in [hunk 11]
```diff
@@ -135,20 +136,48 @@ case "$KEY" in
...
+    fi
+    for i in "${usedSlotsAfterAddKey[@]}"; do
+        if [[ ! " ${usedSlotsBeforeAddKey[@]} " =~ " ${i} " ]]; then
+            SLT=$i
+            break
+        fi
+    done
+fi
+
+if [ -z $SLT ]; then
```
Comment (sergio-correia) [inline src/luks/clevis-luks-bind.in]: Please, double quote `"$SLT"` here.

File: src/luks/clevis-luks-bind.in [hunk 12]
```diff
@@ -135,20 +136,48 @@ case "$KEY" in
...
+    cryptsetup luksAddKey --key-file <(echo -n "${existing_key}") "$DEV" 
 fi < <(echo -n "${key}")
 if [ $? -ne 0 ]; then
     echo "Error while adding new key to LUKS header!" >&2
     exit 1
 fi
 
+#Determine slot used by new key if a desired slot was not specified
+if [ -z $SLT ]; then
+    if [ $luks_type == "luks2" ]; then
```
Comment (sergio-correia) [inline src/luks/clevis-luks-bind.in]: Please, double quote `"$luks_type"` here.

File: src/luks/clevis-luks-bind.in [hunk 13]
```diff
@@ -135,20 +136,48 @@ case "$KEY" in
...
+#Add the new key
 if [ -n "$SLT" ]; then
     cryptsetup luksAddKey --key-slot "$SLT" --key-file \
         <(echo -n "$existing_key") "$DEV"
 else
-    SLT="$(cryptsetup luksAddKey \
-        --key-file <(echo -n "${existing_key}") -v "$DEV" \
-        | sed -rn 's|^Key slot ([0-9]+) created\.$|\1|p')"
+    if [ $luks_type == "luks2" ]; then
+        readarray -t usedSlotsBeforeAddKey < <(cryptsetup luksDump "${DEV}" \
```
[the remark this answers is not in the document]
Comment (sergio-correia) [reply src/luks/clevis-luks-bind.in]: This is important, as it is causing the next line to not parse correctly and causing issues with LUKS2: https://travis-ci.org/sergio-correia/clevis/builds/619890739?utm_source=github_status&utm_medium=notification

File: src/luks/clevis-luks-bind.in [hunk 14]
```diff
@@ -135,20 +136,48 @@ case "$KEY" in
...
-        --key-file <(echo -n "${existing_key}") -v "$DEV" \
-        | sed -rn 's|^Key slot ([0-9]+) created\.$|\1|p')"
+    if [ $luks_type == "luks2" ]; then
+        readarray -t usedSlotsBeforeAddKey < <(cryptsetup luksDump "${DEV}" \ 
+            | sed -rn 's|^\s+([0-9]+): luks2$|\1|p')
+    else
+        readarray -t usedSlotsBeforeAddKey < <(cryptsetup luksDump "${DEV}" \
+            | sed -rn 's|^Key Slot ([0-7]): ENABLED$|\1|p')
+    fi
+    cryptsetup luksAddKey --key-file <(echo -n "${existing_key}") "$DEV"
```
[the remark this answers is not in the document]
Comment (sergio-correia) [reply src/luks/clevis-luks-bind.in]: You forgot to remove the trailing whitespace here. Please remove it as well.

File: src/luks/clevis-luks-bind.in (hunk 9 above)
Comment (sergio-correia) [reply src/luks/clevis-luks-bind.in]: Please, change the tabs in this block to spaces.

**Comment by therealjuanmartinez** [comment]
Feedback appreciated

**Comment by sergio-correia** [comment]
Nice, the build is now green: https://travis-ci.org/sergio-correia/clevis/builds/620254759?utm_medium=notification&utm_source=github_status

I will merge this and update minor styling later. Thanks for the work in this PR!
